[enhancement](cloud) batching get visible version from MetaService

freemandealer opened the pull request:
Get visible versions one by one from MetaService is costly. Batching them into one RPC will not only reduce the workload of RPC service but also reduce the lag.

freemandealer commented:
run buildall

gavinchou commented on fe/fe-core/src/main/java/org/apache/doris/common/NereidsSqlCacheManager.java:
```diff
@@ -253,16 +257,28 @@ private boolean tablesOrDataChanged(Env env, SqlCacheContext sqlCacheContext) {
             OlapTable olapTable = (OlapTable) tableIf;
-            long currentTableTime = olapTable.getVisibleVersionTime();
-            long cacheTableTime = scanTable.latestTimestamp;
+            //long currentTableTime = olapTable.getVisibleVersionTime();
```
revmove unused code

freemandealer replied:
ok, will do after all these p0 cases passes.

gavinchou commented on gensrc/proto/cloud.proto:
```diff
@@ -662,6 +663,7 @@ message CommitTxnResponse {
     repeated int64 partition_ids = 4;
     repeated int64 versions = 5;
     repeated TableStatsPB table_stats = 6;
+    optional int64 version_update_time = 7;
```
What is this for? Why is version_update_time related to RecycleTxnPB

freemandealer replied:
They are for batch and single respectively. What makes you think the relation ever existed?

gavinchou commented on gensrc/proto/cloud.proto:
```diff
@@ -759,6 +761,8 @@ message GetVersionResponse {
     repeated int64 table_ids = 4;
     repeated int64 partition_ids = 5;
     repeated int64 versions = 6;
+    repeated int64 version_update_times = 7;
```
what's the single and multiple update_time for?

freemandealer replied:
ditto

gavinchou commented on cloud/src/meta-service/meta_service_txn.cpp:
```diff
@@ -1043,10 +1043,14 @@ void MetaServiceImpl::commit_txn(::google::protobuf::RpcController* controller,
     }
 
     // Save versions
+    int64_t version_update_time =
```
add time unit as suffix 
e.g. `version_update_time_ms`

pls. check all the other occurrences

freemandealer replied:
sure, will do

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/common/NereidsSqlCacheManager.java:
```diff
@@ -253,16 +257,28 @@ private boolean tablesOrDataChanged(Env env, SqlCacheContext sqlCacheContext) {
             }
 
+            if (Config.isCloudMode()) {
+                Collection<Long> partitionIds = scanTable.getScanPartitions();
+                List<CloudPartition> partitions = partitionIds.stream()
+                        .sorted()
+                        .map(olapTable::getPartition)
+                        .map(partition -> (CloudPartition) partition)
+                        .collect(Collectors.toList());
+                try {
+                    CloudPartition.getSnapshotVisibleVersion(partitions);
+                } catch (RpcException e) {
+                    throw new RuntimeException(e);
+                }
```
add comment to explain what this code block use for

freemandealer replied:
sure, will do

morrySnow commented on fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:
```diff
@@ -697,15 +700,29 @@ private CacheTable buildCacheTableForOlapScanNode(OlapScanNode node) {
+                olapTable.getVisibleVersion());
+        // olapTable.getVisibleVersionTime(), olapTable.getVisibleVersion());
         scanTables.add(scanTable);
+        if (Config.isCloudMode()) {
+            Collection<Long> partitionIds = node.getSelectedPartitionIds();
+            List<CloudPartition> partitions = partitionIds.stream()
+                    .sorted()
+                    .map(olapTable::getPartition)
+                    .map(partition -> (CloudPartition) partition)
+                    .collect(Collectors.toList());
+            try {
+                CloudPartition.getSnapshotVisibleVersion(partitions);
+            } catch (RpcException e) {
+                throw new RuntimeException(e);
+            }
+        }
```
add comment to explain it

w41ter commented on cloud/src/meta-service/meta_service.cpp:
```diff
@@ -261,6 +261,7 @@ void MetaServiceImpl::get_version(::google::protobuf::RpcController* controller,
                 return;
             }
             response->set_version(version);
+            // TODO response->set_version_update_time_ms();
```
```suggestion
```

w41ter commented on cloud/src/meta-service/meta_service.cpp:
```diff
@@ -381,6 +383,7 @@ void MetaServiceImpl::batch_get_version(::google::protobuf::RpcController* contr
                         break;
                     }
                     response->add_versions(version);
+                    // TODO response->add_version_update_times_ms();
```
```suggestion
```

w41ter commented on fe/fe-core/src/main/java/org/apache/doris/qe/cache/CacheAnalyzer.java:
```diff
@@ -697,15 +700,29 @@ private CacheTable buildCacheTableForOlapScanNode(OlapScanNode node) {
+                olapTable.getVisibleVersion());
         scanTables.add(scanTable);
+        if (Config.isCloudMode()) {
+            // get version of related partitions in batch
```
How about extracting this operation into `OlapTable`, just like `OlapTable.selectNonEmptyPartitionIds`?

freemandealer replied:
Ops! On second thought, I think can give it a try.

freemandealer commented:
run buildall

gavinchou reviewed (commented):
LGTM\

freemandealer commented:
run buildall

dataroaring reviewed (dismissed):
LGTM

freemandealer committed:
[enhancement](cloud) batching get visible version from MetaService

Get visible version one by one from MetaService is costly. Batching them
into one RPC will not only reduce the work load of RPC service but also
reduce the lag.

Signed-off-by: freemandealer <freeman.zhang1992@gmail.com>

freemandealer commented:
run buildall

924060929 reviewed (dismissed):
LGTM

freemandealer commented:
run cloud_p1

freemandealer commented:
run buildall

freemandealer committed:
fix cloud_p1

Signed-off-by: freemandealer <freeman.zhang1992@gmail.com>

freemandealer commented:
run buildall

gavinchou commented on gensrc/proto/cloud.proto:
```diff
@@ -818,6 +820,8 @@ message GetVersionResponse {
     repeated int64 partition_ids = 5;
     repeated int64 versions = 6;
+    repeated int64 version_update_times_ms = 7; // for batch
+    optional int64 version_update_time_ms = 8; // for single
```
Consider using only `version_update_times_ms`

freemandealer committed:
response to the reviewer

Signed-off-by: freemandealer <freeman.zhang1992@gmail.com>

freemandealer commented:
run buildall

freemandealer commented:
run p0

freemandealer commented:
run p0

freemandealer commented:
run p0

sollhui reviewed (approved):
LGTM
